app/uk/gov/hmrc/incometaxsessiondata/services/SessionService.scala

```diff
@@ -16,30 +16,107 @@
 
 package uk.gov.hmrc.incometaxsessiondata.services
 
-import uk.gov.hmrc.incometaxsessiondata.models.{Session, SessionData}
+import play.api.Logging
+import play.api.mvc.Result
+import play.api.mvc.Results.{Conflict, Forbidden, InternalServerError, Ok}
+import uk.gov.hmrc.incometaxsessiondata.models.{FullDuplicate, NonDuplicate, PartialDuplicate, Session, SessionData, SessionDataRequest, SessionDuplicationType}
 import uk.gov.hmrc.incometaxsessiondata.repositories.SessionDataRepository
 
 import javax.inject.{Inject, Singleton}
 import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success, Try}
 
 @Singleton
 class SessionService @Inject() (
-  repository: SessionDataRepository
-)(implicit ec: ExecutionContext) {
+  val repository: SessionDataRepository
+)(implicit ec: ExecutionContext) extends Logging {
 
-  def get(sessionId: String): Future[Either[Throwable, Option[SessionData]]] =
-    repository.get(sessionId) map {
+  def get(request: SessionDataRequest[_]): Future[Option[SessionData]] = {
+    repository.get(request) map {
       case Some(data: Session) =>
-        Right(Some(data))
-      case None                => Right(None)
+        Some(SessionData.fromSession(data))
+      case None => None
     }
+  }
 
-  def set(sessionData: SessionData): Future[Boolean] =
-    repository.set(sessionData)
+  def set(session: Session): Future[Either[Throwable, Boolean]] = {
+    Try {
+      repository.set(session)
+    } match {
+      case Failure(exception) => Future.successful(Left(exception))
+      case Success(value) => value.flatMap(res => Future.successful(Right(res.wasAcknowledged)))
+    }
+  }
+
+  def getDuplicationStatus(validRequest: Session): Future[SessionDuplicationType] = {
```

Comment:
we can refactor this later to be Future[Option[SessionDuplicationType]], but we can leave it as is for now